File: mycute-go/src/enum/mode/mode.go
package mode

import (
	"fmt"
	"strings"

	"github.com/t-kawata/mycute/config"
	"github.com/thoas/go-funk"
)

type mode struct {
	value       string
	description string
}

var (
	RT = mode{value: "rt", description: "Run as REST API server."}
	AM = mode{value: "am", description: "Run auto migration for db."}
)

func (m *mode) Val() string {
	return m.value
}

func List() []mode {
	return []mode{RT, AM}
}

func Help() string {
	rtn := []string{fmt.Sprintf("[Available Modes] %s", config.VERSION)}
	for _, m := range List() {
		rtn = append(rtn, fmt.Sprintf("> %s\n\t%s", m.value, m.description))
	}
	return strings.Join(rtn, "\n") + "\n"
}

func IsValidMode(m *string) bool {
	f := funk.Filter(List(), func(md mode) bool {
		return md.value == *m
	})
	return len(f.([]mode)) > 0
}

File: src-from/enums/mode.rs
use crate::config;
use std::str::FromStr;

pub enum Mode {
    RT,
    AM,
}

impl Mode {
    pub fn as_str(&self) -> &str {
        match self {
            Mode::RT => "rt",
            Mode::AM => "am",
        }
    }
    fn as_help(&self) -> &str {
        match self {
            Mode::RT => "Run as REST API server.",
            Mode::AM => "Run auto migration for db.",
        }
    }
    fn all() -> &'static [Mode] {
        &[Mode::RT, Mode::AM]
    }
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "rt" => Some(Mode::RT),
            "am" => Some(Mode::AM),
            _ => None,
        }
    }
    pub fn help() -> String {
        let mut lines = Vec::new();
        lines.push(format!("\n[Available Modes] {}\n", config::VERSION));
        for mode in Self::all() {
            lines.push(format!("> {}\n\t{}", mode.as_str(), mode.as_help()));
        }
        lines.join("\n") + "\n"
    }
    pub fn is_valid(s: &str) -> bool {
        Self::from_str(s).is_some()
    }
}

#[derive(Debug)]
pub struct ModeParseError;

impl FromStr for Mode {
    type Err = ModeParseError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::from_str(s).ok_or(ModeParseError)
    }
}
